packages/gatsby/cache-dir/ssr-develop-static-entry.js

```diff
@@ -111,7 +115,81 @@ export default (pagePath, isClientOnlyPage, callback) => {
 
     const pageData = getPageData(pagePath)
 
-    const componentChunkName = pageData?.componentChunkName
+    const { componentChunkName, staticQueryHashes = [] } = pageData
+
+    let scriptsAndStyles = flatten(
+      [`commons`].map(chunkKey => {
+        const fetchKey = `assetsByChunkName[${chunkKey}]`
+
+        let chunks = get(stats, fetchKey)
+        const namedChunkGroups = get(stats, `namedChunkGroups`)
+
+        if (!chunks) {
+          return null
+        }
+
+        chunks = chunks.map(chunk => {
+          if (chunk === `/`) {
+            return null
+          }
+          return { rel: `preload`, name: chunk }
+        })
+
+        namedChunkGroups[chunkKey].assets.forEach(asset =>
+          chunks.push({ rel: `preload`, name: asset })
+        )
+
+        const childAssets = namedChunkGroups[chunkKey].childAssets
+        for (const rel in childAssets) {
+          chunks = concat(
+            chunks,
+            childAssets[rel].map(chunk => {
+              return { rel, name: chunk }
+            })
+          )
+        }
+
+        return chunks
+      })
+    )
+      .filter(s => isObject(s))
+      .sort((s1, s2) => (s1.rel == `preload` ? -1 : 1)) // given priority to preload
+
+    scriptsAndStyles = uniqBy(scriptsAndStyles, item => item.name)
+
+    const styles = scriptsAndStyles.filter(
+      style => style.name && style.name.endsWith(`.css`)
+    )
+
+    styles
+      .slice(0)
+      .reverse()
+      .forEach(style => {
+        headComponents.unshift(
+          <script
+            dangerouslySetInnerHTML={{
+              __html: `
+            // Immediately update the styles link with its absolute URL so HMR
+            // will work
+            var scripts = document.getElementsByTagName('script')
+            var currentScript = scripts[scripts.length - 1]
+            var linkEl = currentScript.previousSibling
+            var newHref = new URL("${__PATH_PREFIX__}/${style.name}", location.href).href
+            linkEl.href = newHref
+            `,
+            }}
+          />
+        )
+        headComponents.unshift(
+          <link
+            data-identity={`gatsby-dev-css`}
```

Comment:
this attribute doesn't seem used anywhere? can we drop it?

Reply:
it's used in the ssr tests when we compare the prod & dev ssr — we have to remove specific prod/dev differences so the diff is otherwise the same https://github.com/gatsbyjs/gatsby/blob/3c088a58aca6620cfc3077dbf0af0b771ff45a3a/integration-tests/ssr/test-output.js#L27

Reply:
Can we then use same identifier as in prod then instead?
https://github.com/gatsbyjs/gatsby/blob/23da2c3fb2e16b7e3fe1e15c19decd799000a212/packages/gatsby/cache-dir/static-entry.js#L418

Reply:
yeah let's do that 👍

Reply:
actually... no we can't — in prod there'll only be one component added where here there might be multiple css chunks generated.

Reply:
not sure why the difference tbh

Reply:
Hmm, if we are aiming to use minicssextract to also take care of differences between dev and prod - then I think we should also ensure single css chunk for dev (or rather - same setting for dev as for prod when it comes to chunking)?

Reply:
yeah we'd need to do this for dev as well https://github.com/gatsbyjs/gatsby/blob/6d3a64b2398c9019ee95af5e39a3b41ceab861be/packages/gatsby/src/utils/webpack.config.js#L598-L608

Reply:
btw, that test _is_ the exception for CSS Modules right so we can remove the TODO?

Reply:
I'm not exactly sure what the TODO refers too. @wardpeet will have most context on it

Reply:
or maybe not https://github.com/gatsbyjs/gatsby/blob/6d3a64b2398c9019ee95af5e39a3b41ceab861be/packages/gatsby/src/utils/webpack.config.js#L529

Reply:
hmmm setting `chunks: all` in dev seems to disable all HMR. I thought it was just css but I tried editing a page component and it wouldn't hot reload as all. So not sure if it's possible to enforce moving all styles into one chunk in dev.

Reply:
Does it disable HMR ... or runtime just doesn't work because we chunk things other than styles too and `common.js` is not created anymore?

Reply:
I put `chunks: all` inside the `styles` `cacheGroup` so the commons.js is (should be?) unaffected. It's created like normal and rebuilds like normal e.g. if you refresh changes show up.

Reply:
But if runtime wouldn't work, the refresh with dev ssr would paint same page, right? :) That's why I asked

Reply:
This is the diff if you want to play with it:

```diff
diff --git a/packages/gatsby/src/utils/webpack.config.js b/packages/gatsby/src/utils/webpack.config.js
index 6c5c68303f..3a616ab5da 100644
--- a/packages/gatsby/src/utils/webpack.config.js
+++ b/packages/gatsby/src/utils/webpack.config.js
@@ -700,6 +700,27 @@ module.exports = async (
     config.externals = {
       "socket.io-client": `io`,
     }
+
+    const isCssModule = module => module.type === `css/mini-extract`
+    config.optimization = {
+      splitChunks: {
+        cacheGroups: {
+          // Bundle all css & lazy css into one stylesheet to make sure lazy components do not break
+          // TODO make an exception for css-modules
+          styles: {
+            test(module) {
+              return isCssModule(module)
+            },
+
+            chunks: `all`,
+            name: `styles`,
+            priority: 40,
+            enforce: true,
+          },
+        },
+      },
+    }
   }
 
   store.dispatch(actions.replaceWebpackConfig(config))

```

Reply:
>But if runtime wouldn't work, the refresh with dev ssr would paint same page, right? :) That's why I asked

Not sure I understand? The compilation is still happening — it's just not getting hot reloaded. If you do a full refresh you see your changes.

Reply:
> Not sure I understand? The compilation is still happening — it's just not getting hot reloaded. If you do a full refresh you see your changes.

Does react runtime actually mounts? When I was previously checking chunking for dev, the problem I saw was that runtime actually didn't work / react CLIENT application didn't mount (without actual errors). If this is same issue I was seeing - then dev ssr can "hide" this problem heh, by showing usable html (just not interactive one). And then HMR not working also make sense because nothing is even mounted

Reply:
Ok, so anyway - let's keep it as is then for now ;) I guess we can try to address chunking separately some time later as it seems much more involved than initially seemed